code_mode_tools panic

```txt
panic: runtime error: invalid memory address or nil pointer dereference [recovered, repanicked]
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x23b14ef]

goroutine 47 [running]:
github.com/dop251/goja.(*Runtime).RunProgram.func1()
	/root/go/pkg/mod/github.com/dop251/goja@v0.0.0-20260311135729-065cd970411c/runtime.go:1454 +0x2aa
panic({0x2ba52c0?, 0x5893f80?})
	/snap/go/11127/src/runtime/panic.go:860 +0x13a
github.com/dop251/goja.(*vm).handleThrow(0x14a88a4d6d80, {0x2ba52c0, 0x5893f80})
	/root/go/pkg/mod/github.com/dop251/goja@v0.0.0-20260311135729-065cd970411c/vm.go:841 +0x425
github.com/dop251/goja.(*vm).runTryInner.func1()
	/root/go/pkg/mod/github.com/dop251/goja@v0.0.0-20260311135729-065cd970411c/vm.go:883 +0x3a
panic({0x2ba52c0?, 0x5893f80?})
	/snap/go/11127/src/runtime/panic.go:860 +0x13a
github.com/docker/docker-agent/pkg/tools/codemode.(*codeModeTool).runJavascript.callTool.func1(0x14a889ea0330)
	/root/docker-agent/pkg/tools/codemode/exec.go:122 +0x30f
reflect.Value.call({0x2b26840?, 0x14a889f90900?, 0x14a88a2cae88?}, {0x328ab89, 0x4}, {0x14a88a2cafe8, 0x1, 0x14a88a2caeb0?})
	/snap/go/11127/src/reflect/value.go:586 +0xf06
reflect.Value.Call({0x2b26840?, 0x14a889f90900?, 0x14a889ea00c0?}, {0x14a88a2cafe8?, 0x14a889a022b0?, 0x1400a88a2cafe8?})
	/snap/go/11127/src/reflect/value.go:369 +0xb9
github.com/dop251/goja.(*Runtime).newWrappedFunc.(*Runtime).wrapReflectFunc.func1({{0x36738e0, 0x5be6940}, {0x14a88ab7a1c0, 0x1, 0x4}})
	/root/go/pkg/mod/github.com/dop251/goja@v0.0.0-20260311135729-065cd970411c/runtime.go:2035 +0x405
github.com/dop251/goja.(*nativeFuncObject).vmCall(0x14a889a87740, 0x14a88a4d6d80, 0x1)
	/root/go/pkg/mod/github.com/dop251/goja@v0.0.0-20260311135729-065cd970411c/func.go:571 +0x184
github.com/dop251/goja.call.exec(0xf?, 0x14a88a4d6d80)
	/root/go/pkg/mod/github.com/dop251/goja@v0.0.0-20260311135729-065cd970411c/vm.go:3642 +0x66
github.com/dop251/goja.(*vm).run(0x14a88a4d6d80)
	/root/go/pkg/mod/github.com/dop251/goja@v0.0.0-20260311135729-065cd970411c/vm.go:635 +0x5b
github.com/dop251/goja.(*vm).runTryInner(0x30acbe0?)
	/root/go/pkg/mod/github.com/dop251/goja@v0.0.0-20260311135729-065cd970411c/vm.go:887 +0x52
github.com/dop251/goja.(*vm).runTry(0x14a88a4d6d80)
	/root/go/pkg/mod/github.com/dop251/goja@v0.0.0-20260311135729-065cd970411c/vm.go:873 +0x1b4
github.com/dop251/goja.(*Runtime).RunProgram(0x14a88a040c08, 0x14a88a7b21e0)
	/root/go/pkg/mod/github.com/dop251/goja@v0.0.0-20260311135729-065cd970411c/runtime.go:1476 +0x425
github.com/dop251/goja.(*Runtime).RunScript(0x14a88a040c08, {0x0?, 0x3607a40?}, {0x14a88a7d0000?, 0x14a889f90900?})
	/root/go/pkg/mod/github.com/dop251/goja@v0.0.0-20260311135729-065cd970411c/runtime.go:1412 +0x50
github.com/dop251/goja.(*Runtime).RunString(...)
	/root/go/pkg/mod/github.com/dop251/goja@v0.0.0-20260311135729-065cd970411c/runtime.go:1401
github.com/docker/docker-agent/pkg/tools/codemode.(*codeModeTool).runJavascript(0x14a889e9c6c0, {0x363e1c8, 0x14a889d38150}, {0x14a889ffab00, 0x15d})
	/root/docker-agent/pkg/tools/codemode/exec.go:66 +0x40c
github.com/docker/docker-agent/pkg/tools/codemode.(*codeModeTool).Tools.func1({0x363e1c8?, 0x14a889d38150?}, {{0x14a889ffab00?, 0x298b9a0?}})
	/root/docker-agent/pkg/tools/codemode/codemode.go:82 +0x4a
github.com/docker/docker-agent/pkg/tools/codemode.(*codeModeTool).Tools.NewHandler[...].func2({{0x14a889ac4997, 0x1d}, {0x14a88a981968, 0x8}, {{0x14a88a980198, 0x19}, {0x14a88a738600, 0x16a}}})
	/root/docker-agent/pkg/tools/tools.go:27 +0xc4
github.com/docker/docker-agent/pkg/runtime.(*LocalRuntime).runTool.func1({0x363e1c8?, 0x14a889d38150?})
	/root/docker-agent/pkg/runtime/tool_dispatch.go:342 +0x55
github.com/docker/docker-agent/pkg/runtime.(*LocalRuntime).executeToolWithHandler(0x14a88a29afc0, {0x363e1c8, _}, {{0x14a889ac4997, 0x1d}, {0x14a88a981968, 0x8}, {{0x14a88a980198, 0x19}, {0x14a88a738600, ...}}}, ...)
	/root/docker-agent/pkg/runtime/tool_dispatch.go:268 +0x92f
github.com/docker/docker-agent/pkg/runtime.(*LocalRuntime).runTool(0x14a88a29afc0, {0x363e1c8, _}, {{0x32cbf6c, 0x19}, {0x32983bf, 0x9}, {0x14a88a7ea800, 0x4367}, {0x3096ec0, ...}, ...}, ...)
	/root/docker-agent/pkg/runtime/tool_dispatch.go:340 +0x34c
github.com/docker/docker-agent/pkg/runtime.(*LocalRuntime).processToolCalls.func2()
	/root/docker-agent/pkg/runtime/tool_dispatch.go:69 +0xe5
github.com/docker/docker-agent/pkg/runtime.(*LocalRuntime).executeWithApproval(0x0?, {0x363e1c8?, _}, _, {{0x14a889ac4997, 0x1d}, {0x14a88a981968, 0x8}, {{0x14a88a980198, 0x19}, ...}}, ...)
	/root/docker-agent/pkg/runtime/tool_dispatch.go:117 +0x589
github.com/docker/docker-agent/pkg/runtime.(*LocalRuntime).processToolCalls(0x14a88a29afc0, {0x363e1c8, 0x14a88a2a1800}, 0x14a889905080, {0x14a889d44140, 0x1, 0x1}, {0x14a889a2e008, 0x6, 0xb}, ...)
	/root/docker-agent/pkg/runtime/tool_dispatch.go:73 +0x1058
github.com/docker/docker-agent/pkg/runtime.(*LocalRuntime).RunStream.func1()
	/root/docker-agent/pkg/runtime/loop.go:383 +0x3238
created by github.com/docker/docker-agent/pkg/runtime.(*LocalRuntime).RunStream in goroutine 1
	/root/docker-agent/pkg/runtime/loop.go:77 +0x1a9 
```

--- a/pkg/toolinstall/resolver.go
+++ b/pkg/toolinstall/resolver.go
@@ -25,23 +25,14 @@ func EnsureCommand(ctx context.Context, command, version string) (string, error)
 		return command, nil
 	}
 
-	version = strings.TrimSpace(version)
-	lower := strings.ToLower(version)
+	lower := strings.ToLower(strings.TrimSpace(version))
 	if lower == "false" || lower == "off" {
 		return command, nil
 	}
 
 	resolvedPath, err := resolve(ctx, command, version)
 	if err != nil {
-		if version != "" {
-			return "", fmt.Errorf("auto-installing command %q: %w", command, err)
-		}
-
-		slog.Warn("Auto-install failed, falling back to original command",
-			"command", command,
-			"error", err,
-		)
-		return command, nil
+		return "", fmt.Errorf("auto-installing command %q: %w", command, err)
 	}
 
 	return resolvedPath, nil

--- a/pkg/toolinstall/resolver_test.go
+++ b/pkg/toolinstall/resolver_test.go
@@ -50,16 +50,7 @@ func TestEnsureCommand_DisabledPerToolset(t *testing.T) {
 	}
 }
 
-func TestEnsureCommand_AutoDetectFailureFallsBackToOriginalCommand(t *testing.T) {
-	t.Setenv("DOCKER_AGENT_TOOLS_DIR", t.TempDir())
-	t.Setenv("DOCKER_AGENT_AUTO_INSTALL", "")
-
-	result, err := EnsureCommand(t.Context(), "nonexistent-tool", "")
-	require.NoError(t, err)
-	assert.Equal(t, "nonexistent-tool", result)
-}
-
-func TestEnsureCommand_ExplicitVersionFailureStillErrors(t *testing.T) {
+func TestEnsureCommand_SoftFail(t *testing.T) {
 	t.Setenv("DOCKER_AGENT_TOOLS_DIR", t.TempDir())
 	t.Setenv("DOCKER_AGENT_AUTO_INSTALL", "")
 
@@ -82,7 +73,7 @@ func TestEnsureCommand_FoundInBinDir(t *testing.T) {
 	assert.Equal(t, fakeBin, result)
 }
 
-func TestEnsureCommand_NonExecutableInBinDirFallsBackToOriginalCommand(t *testing.T) {
+func TestEnsureCommand_NonExecutableInBinDirIsSkipped(t *testing.T) {
 	toolsDir := t.TempDir()
 	t.Setenv("DOCKER_AGENT_TOOLS_DIR", toolsDir)
 	t.Setenv("DOCKER_AGENT_AUTO_INSTALL", "")
@@ -91,9 +82,9 @@ func TestEnsureCommand_NonExecutableInBinDirFallsBackToOriginalCommand(t *testin
 	require.NoError(t, os.MkdirAll(binDir, 0o755))
 	require.NoError(t, os.WriteFile(filepath.Join(binDir, "not-executable"), []byte("data"), 0o644))
 
-	result, err := EnsureCommand(t.Context(), "not-executable", "")
-	require.NoError(t, err)
-	assert.Equal(t, "not-executable", result)
+	// Falls through to auto-install → fails → returns error.
+	_, err := EnsureCommand(t.Context(), "not-executable", "")
+	require.Error(t, err)
 }
 
 // --- resolve tests ---

--- a/pkg/tools/codemode/exec.go
+++ b/pkg/tools/codemode/exec.go
@@ -89,57 +89,62 @@ func (c *codeModeTool) runJavascript(ctx context.Context, script string) (Script
 
 func callTool(ctx context.Context, tool tools.Tool, tracker *toolCallTracker) func(args map[string]any) (string, error) {
 	return func(args map[string]any) (string, error) {
-		var toolArgs struct {
-			Required []string `json:"required"`
-		}
+		output, filtered, err := invokeTool(ctx, tool, args)
 
-		if err := tools.ConvertSchema(tool.Parameters, &toolArgs); err != nil {
-			tracker.record(ToolCallInfo{
-				Name:      tool.Name,
-				Arguments: args,
-				Error:     err.Error(),
-			})
-			return "", err
-		}
-
-		nonNilArgs := make(map[string]any)
-		for k, v := range args {
-			if slices.Contains(toolArgs.Required, k) || v != nil {
-				nonNilArgs[k] = v
-			}
-		}
-
-		arguments, err := json.Marshal(nonNilArgs)
-		if err != nil {
-			tracker.record(ToolCallInfo{
-				Name:      tool.Name,
-				Arguments: nonNilArgs,
-				Error:     err.Error(),
-			})
-			return "", err
-		}
-
-		result, err := tool.Handler(ctx, tools.ToolCall{
-			Function: tools.FunctionCall{
-				Name:      tool.Name,
-				Arguments: string(arguments),
-			},
-		})
-		if err != nil {
-			tracker.record(ToolCallInfo{
-				Name:      tool.Name,
-				Arguments: nonNilArgs,
-				Error:     err.Error(),
-			})
-			return "", err
-		}
-
-		tracker.record(ToolCallInfo{
+		info := ToolCallInfo{
 			Name:      tool.Name,
-			Arguments: nonNilArgs,
-			Result:    result.Output,
-		})
+			Arguments: filtered,
+		}
+		if err != nil {
+			info.Error = err.Error()
+		} else {
+			info.Result = output
+		}
+		tracker.record(info)
 
-		return result.Output, nil
+		return output, err
 	}
+}
+
+// invokeTool calls a single tool handler, filtering out nil optional arguments.
+// It returns the output, the filtered arguments actually sent, and any error.
+func invokeTool(ctx context.Context, tool tools.Tool, args map[string]any) (string, map[string]any, error) {
+	if tool.Handler == nil {
+		return "", args, fmt.Errorf("tool %q is not available in code mode", tool.Name)
+	}
+
+	var schema struct {
+		Required []string `json:"required"`
+	}
+	if err := tools.ConvertSchema(tool.Parameters, &schema); err != nil {
+		return "", args, err
+	}
+
+	// Strip nil optional arguments that goja passes for omitted parameters.
+	filtered := make(map[string]any)
+	for k, v := range args {
+		if slices.Contains(schema.Required, k) || v != nil {
+			filtered[k] = v
+		}
+	}
+
+	arguments, err := json.Marshal(filtered)
+	if err != nil {
+		return "", filtered, err
+	}
+
+	result, err := tool.Handler(ctx, tools.ToolCall{
+		Function: tools.FunctionCall{
+			Name:      tool.Name,
+			Arguments: string(arguments),
+		},
+	})
+	if err != nil {
+		return "", filtered, err
+	}
+
+	if result == nil {
+		return "", filtered, nil
+	}
+	return result.Output, filtered, nil
 }

